### DISCUSSION TITLE
About the BlenderStyle

### DISCUSSION THREAD

**Discussion by ZiguoAtGitHub** [marcomusy/vedo#904] [Q&A] [answered]
Hi Marco,
I noticed the new functionality of interaction mode and took a try on the example. 
However I got nothing selected inside the rectangle zone generated  by left button of mouse.
Did I miss something or there is no select functionality in the BlenderStyle?
To be honest, I have studied all the codes of BlenderStyle but haven't made it clear yet.

**Comment by marcomusy** [comment]
Honestly I don't know the answer to that! @RubendeBruin

**Reply by RubendeBruin** [reply]
Hi @ZiguoAtGitHub, can you paste the code that you used than I can have a look.

**Reply by ZiguoAtGitHub** [reply]
Hi you all @RubendeBruin  @marcomusy ,
Here is the code. In fact, It is just copied from vedo's api documentation.🤦‍♂️
```
from vedo import *
settings.enable_default_keyboard_callbacks = False
settings.enable_default_mouse_callbacks = False
mesh = Mesh(dataurl+"cow.vtk")
mode = interactor_modes.BlenderStyle()
plt = Plotter().user_mode(mode)
plt.show(mesh, axes=1)
```
And maybe I should make what I want and my problem more clearly:
I want to select all cells inside one rectangle zone which is created under BlenderStyle mode.
And I thought it could be done automatically under BlenderStyle mode.
![select](https://github.com/marcomusy/vedo/assets/116484589/854eac3f-bd20-4475-bbee-8dfd2d9cb2a1)

**Reply by ZiguoAtGitHub** [reply]
If there is no such functionality, how to add it? I suppose two steps should be taken:
1) Project all cells' vertex to screen and
2) Select those inside the rectangle.
Are there some other better ways or some convenient functions already provided by vedo?
Thanks for any suggestions.

**Comment by RubendeBruin** [comment]
Hi @ZiguoAtGitHub , at the moment the interactor uses the PickProp() method of the vtk renderer. This only returns "props" (actors).

https://vtk.org/doc/nightly/html/classvtkRenderer.html#ae7fce027da043f2c5ebf4fb0ebb8da15

A quick-fix could be adding small cubes at all the vertices. Then those will be returned as selection. But will not be quick for large models.

**Reply by ZiguoAtGitHub** [reply]
Hi @RubendeBruin ，
Thank you very much!!
It is no very good way in VTK by now under my understanding toward your replies, isn't it?
Maybe I need dig the documents of vedo and VTK deeper to get what I want.

**Reply by marcomusy** [reply, marked as answer]
This seems to go in the right direction:

```py
from vedo import *
import vtk

settings.enable_default_mouse_callbacks = False
# settings.use_parallel_projection = True


def mode_select(objs):

    d0 = mode.start_x, mode.start_y
    d1 = mode.end_x, mode.end_y
    
    area_picker = vtk.vtkAreaPicker()
    area_picker.AreaPick(*d0, *d1, plt.renderer)
    planes = area_picker.GetFrustum()

    fru = vtk.vtkFrustumSource()
    fru.SetPlanes(planes)
    fru.ShowLinesOff()
    fru.Update()

    afru = Mesh(fru.GetOutput())
    afru.alpha(0.2).lw(1).pickable(False)

    infru = afru.inside_points(mesh)
    infru.ps(10).c('purple7')

    p0 = plt.compute_world_coordinate(d0)
    p1 = plt.compute_world_coordinate(d1)

    plt.add(afru, infru, Point(p0), Point(p1))
    plt.render()

mesh = Mesh(dataurl+"cow.vtk").lw(1)

mode = interactor_modes.BlenderStyle()
mode.callbackSelect = mode_select

plt = Plotter().user_mode(mode)
plt.show(mesh, axes=1)
```

![Screenshot from 2023-07-25 21-37-07](https://github.com/marcomusy/vedo/assets/32848391/203cbc4b-caf5-4aa7-ad6c-ae41c514af02)

I will try to make its interface more user-friendly..

**Reply by ZiguoAtGitHub** [reply]
Wa Oh! I can't say anything but AWSOME!
Thank you Marco! That script almost give me the answer and I think I can make the last step on your instructive work.
👍👍👍

**Reply by marcomusy** [reply]
This is a more compact - hopefully more intuitive - version of the API:
```
pip install -U git+https://github.com/marcomusy/vedo.git
```

then:


```py
from vedo import *

settings.enable_default_mouse_callbacks = False

def mode_select(objs):
    print("Selected objects:", objs)
    d0 = mode.start_x, mode.start_y  # display coords
    d1 = mode.end_x,   mode.end_y

    frustum = plt.pick_area(d0, d1)
    infru = frustum.inside_points(mesh, return_ids=False)
    color = np.random.randint(0, 10)
    infru.ps(10).c(color)
    plt.add(frustum, infru).render()

mesh = Mesh(dataurl+"cow.vtk").c("k5").lw(1)

mode = interactor_modes.BlenderStyle()
mode.callback_select = mode_select

plt = Plotter()
plt.show(mesh, axes=1, mode=mode)
```
![Screenshot from 2023-07-26 14-00-29](https://github.com/marcomusy/vedo/assets/32848391/b185495a-22e4-4f10-b95b-55d08848092d)

note that you can play with `return_ids` to access the vertices of the mesh.

**Reply by ZiguoAtGitHub** [reply]
Hi @marcomusy ,
 I met one problem today when I tried to add callbacks based on mode 6. The callbacks are both related to left button but response to right button. Here is the code:
```
from vedo import *
import vtk

settings.enable_default_mouse_callbacks = False
# settings.use_parallel_projection = True

def LButtonPress(evt):
  plt.start_x,plt.start_y=evt.picked2d
  return

def LButtonUp(evt):
  plt.end_x,plt.end_y=evt.picked2d

  d0 = plt.start_x,plt.start_y
  d1 = plt.end_x,plt.end_y

  if ((plt.end_x-plt.start_x)**2+(plt.end_y-plt.start_y)**2)**0.5>100:
    area_picker = vtk.vtkAreaPicker()
    area_picker.AreaPick(*d0, *d1, plt.renderer)
    planes = area_picker.GetFrustum()
    
    fru = vtk.vtkFrustumSource()
    fru.SetPlanes(planes)
    fru.ShowLinesOff()
    fru.Update()

    afru = Mesh(fru.GetOutput())
    afru.alpha(0.2).lw(1).pickable(False)

    infru = afru.inside_points(mesh)
    infru.ps(10).c('purple7')

    p0 = plt.compute_world_coordinate(d0)
    p1 = plt.compute_world_coordinate(d1)

    plt.add(afru, infru, Point(p0), Point(p1))
    plt.render()
  return

mesh = Mesh(dataurl+"cow.vtk").lw(1)

plt = Plotter()

plt.start_x,plt.start_y=0.0,0.0
plt.end_x,plt.end_y=0.0,0.0
plt.add_callback('LeftButtonPress',LButtonPress)
plt.add_callback('LeftButtonRelease',LButtonUp)

plt.show(mesh, axes=1,mode=6)

```
Thank you very much! And your work are so much awsome!

**Reply by ZiguoAtGitHub** [reply]
extra-information, the callbacks response to middle button also.😂

**Comment by marcomusy** [comment]
Hi @ZiguoAtGitHub sorry for the late reply i was on holiday :)
After 
```
pip install -U git+https://github.com/marcomusy/vedo.git
```

try
```py
from vedo import *

settings.enable_default_mouse_callbacks = False


def LButtonPress(evt):
    print("LButtonPress")
    plt.start_x, plt.start_y = evt.picked2d
    plt.render()


def LButtonUp(evt):
    print("LButtonUp")
    plt.end_x, plt.end_y = evt.picked2d

    d0 = np.array([plt.start_x, plt.start_y])
    d1 = evt.picked2d
    
    if mag(d0-d1) > 100:
        frustum = plt.pick_area(d0, d1)
        infru = frustum.inside_points(mesh).ps(10).c("purple7")
        plt.add(frustum, infru)
    plt.render()


mesh = Mesh(dataurl + "cow.vtk").lw(1)

plt = Plotter()
plt.start_x, plt.start_y = 0.0, 0.0
plt.add_callback("left button press", LButtonPress)
plt.add_callback("left button release", LButtonUp)
plt.show(mesh, axes=1, mode=6)
```

**Comment by ZiguoAtGitHub** [comment]
Hi @marcomusy 
Thanks! I tried and got what I want.
It must be a wonderful holiday you just enjoyed!  😁
